pkg/cmd/util/tokencmd/negotiate_gssapi.go

```diff
@@ -0,0 +1,211 @@
+	// TODO: surface option for mutual authentication, e.g. gssapi.GSS_C_MUTUAL_FLAG
+	flags uint32
+
+	// tracks whether the managed resources have been released
+	released bool
+	// tracks whether gssapi requires more calls to InitSecContext before considering the negotiation complete
+	continueNeeded bool
+}
+
+func NewNegotiateChallengeHandler() ChallengeHandler {
+	return &NegotiateChallengeHandler{}
+}
+
+func (c *NegotiateChallengeHandler) CanHandle(headers http.Header) bool {
+	// Make sure this is a negotiate request
+	isNegotiate, _, err := getNegotiateToken(headers)
+	if err != nil || !isNegotiate {
+		return false
+	}
+	// Make sure we can get a gssapi library
+	_, err = c.loadLib()
+	return err == nil
+}
+
+func (c *NegotiateChallengeHandler) HandleChallenge(requestURL string, headers http.Header) (http.Header, bool, error) {
+	if c.released {
+		glog.V(5).Infof("already completed negotiation, cannot start another one")
+		return nil, false, errors.New("already completed negotiation, cannot start another one")
+	}
+
+	lib, err := c.loadLib()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Get incoming token
+	_, incomingToken, err := getNegotiateToken(headers)
+	if err != nil {
+		return nil, false, err
+	}
+	incomingTokenBuffer, err := lib.MakeBufferBytes(incomingToken)
+	if err != nil {
+		return nil, false, err
+	}
+	defer incomingTokenBuffer.Release()
+
+	if c.ctx == nil {
+		u, err := url.Parse(requestURL)
+		if err != nil {
+			return nil, false, err
+		}
+
+		glog.V(5).Infof("importing service name %s", u.Host)
+		nameBuf, err := lib.MakeBufferString("HTTP@" + u.Host)
+		if err != nil {
+			return nil, false, err
+		}
+		defer nameBuf.Release()
+
+		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		if err != nil {
+			return nil, false, err
+		}
+		c.name = name
+		c.ctx = lib.GSS_C_NO_CONTEXT
+	}
+
+	// Call InitSecContext to validate the incoming token (if given), and get our outgoing token (if needed)
+	// Save the returned context
+	// TODO: validate returned flags support requested services
+	var outgoingToken *gssapi.Buffer
+	c.ctx, _, outgoingToken, _, _, err = lib.InitSecContext(lib.GSS_C_NO_CREDENTIAL, c.ctx, c.name, lib.GSS_C_NO_OID, c.flags, time.Duration(0), lib.GSS_C_NO_CHANNEL_BINDINGS, incomingTokenBuffer)
+	defer outgoingToken.Release()
+
+	switch err {
+	case nil:
+		glog.V(5).Infof("InitSecContext returned GSS_S_COMPLETE")
+		c.continueNeeded = false
+	case gssapi.ErrContinueNeeded:
+		glog.V(5).Infof("InitSecContext returned GSS_S_CONTINUE_NEEDED")
+		c.continueNeeded = true
+	default:
+		glog.V(5).Infof("InitSecContext returned error: %v (%#v)", err, err)
+		return nil, false, err
+	}
+
+	responseHeaders := http.Header{}
+	responseHeaders.Set("Authorization", "Negotiate "+base64.StdEncoding.EncodeToString(outgoingToken.Bytes()))
+	return responseHeaders, true, nil
+}
+
+func (c *NegotiateChallengeHandler) CompleteChallenge(requestURL string, headers http.Header) error {
```

Comment:
called once we complete a negotiation (we get a non-401 from the server). gives the client a chance to indicate it is satisfied with the negotiation (in a mutual auth case, for example)